src/qml/qmlpreferencesproxy.cpp

```diff
@@ -0,0 +1,262 @@
+#include "qmlpreferencesproxy.h"
+
+#include <qqmlengine.h>
+
+#include <memory>
+
+#include "moc_qmlpreferencesproxy.cpp"
+#include "soundio/soundmanager.h"
+#include "soundio/soundmanagerutil.h"
+#include "util/assert.h"
+#include "util/scopedoverridecursor.h"
+
+namespace mixxx {
+namespace qml {
+
+namespace {
+const QString kAppGroup = QStringLiteral("[App]");
+const ConfigKey kKeylockEngineCfgkey =
+        ConfigKey(kAppGroup, QStringLiteral("keylock_engine"));
+
+} // namespace
+
+QmlSoundDeviceProxy::QmlSoundDeviceProxy(
+        SoundDevicePointer pInternal, Type type, QObject* parent)
+        : QObject(parent),
+          m_pInternal(pInternal),
+          m_type(type) {
+}
+uint QmlSoundDeviceProxy::getChannelCount() const {
+    return m_type == Type::Input ? m_pInternal->getNumInputChannels()
+                                 : m_pInternal->getNumOutputChannels();
+}
+SoundDeviceId QmlSoundDeviceProxy::getDeviceId() const {
+    return m_pInternal->getDeviceId();
+}
+
+QList<QmlSoundDeviceConnection*> QmlSoundDeviceProxy::connections(QmlSoundManagerProxy* manager) {
+    QList<QmlSoundDeviceConnection*> connections;
+
+    auto pManager = manager->internal();
+    auto config = pManager->getConfig();
+
+    if (m_type == Type::Input) {
```

Comment:
Rather than create one type that can do both input and output, use inheritance:

QmlSoundDeviceConnection should be the base class
QmlSoundDeviceInputConnection (or ConnectionInput)? Can contain specializations for input
QmlSoundDeviceOutputputConnection can contain specializations for output.  

There should be no/minimal code duplication, just a couple of overrides for the parts where input differs from output.